source/server/win/server_transport_winhttp.c

```diff
@@ -34,7 +34,76 @@ static HINTERNET get_winhttp_req(HttpTransportContext *ctx, const char *directio
 		return NULL;
 	}
 
-	if (ctx->proxy_user)
+	// if no proxy is set, we should look to see if we can (and should) use the system
+	// proxy settings for the given user.
+	if (!ctx->proxy)
+	{
+		if (!ctx->proxy_configured)
+		{
+			WINHTTP_CURRENT_USER_IE_PROXY_CONFIG ieConfig = { 0 };
+			if (WinHttpGetIEProxyConfigForCurrentUser(&ieConfig))
+			{
+				dprintf("[PROXY] Got IE configuration");
+				if (ieConfig.fAutoDetect && ieConfig.lpszAutoConfigUrl)
+				{
+					WINHTTP_AUTOPROXY_OPTIONS autoProxyOpts = { 0 };
+					WINHTTP_PROXY_INFO proxyInfo = { 0 };
+
+					dprintf("[PROXY] IE config set to autodetect with URL %S", ieConfig.lpszAutoConfigUrl);
+
+					autoProxyOpts.dwFlags = WINHTTP_AUTOPROXY_AUTO_DETECT | WINHTTP_AUTOPROXY_CONFIG_URL;
+					autoProxyOpts.dwAutoDetectFlags = WINHTTP_AUTO_DETECT_TYPE_DHCP | WINHTTP_AUTO_DETECT_TYPE_DNS_A;
+					autoProxyOpts.fAutoLogonIfChallenged = TRUE;
+					autoProxyOpts.lpszAutoConfigUrl = ieConfig.lpszAutoConfigUrl;
+
+					if (WinHttpGetProxyForUrl(ctx->internet, ctx->url, &autoProxyOpts, &proxyInfo))
+					{
+						ctx->proxy_for_url = malloc(sizeof(WINHTTP_PROXY_INFO));
+						memcpy(ctx->proxy_for_url, &proxyInfo, sizeof(WINHTTP_PROXY_INFO));
+					}
+				}
+				else if (ieConfig.lpszProxy)
+				{
+					WINHTTP_PROXY_INFO* proxyInfo = (WINHTTP_PROXY_INFO*)calloc(1, sizeof(WINHTTP_PROXY_INFO));
+					ctx->proxy_for_url = proxyInfo;
+
+					dprintf("[PROXY] IE config set to proxy %S with bypass %S", ieConfig.lpszProxy, ieConfig.lpszProxyBypass);
+
+					proxyInfo->dwAccessType = WINHTTP_ACCESS_TYPE_NAMED_PROXY;
+					proxyInfo->lpszProxy = ieConfig.lpszProxy;
+					proxyInfo->lpszProxyBypass = ieConfig.lpszProxyBypass;
+
+					// stop the cleanup code from removing these as we're using them behind the scenes and they will
+					// be freed later instead.
+					ieConfig.lpszProxy = NULL;
+					ieConfig.lpszProxyBypass = NULL;;
+				}
+
+				if (ieConfig.lpszAutoConfigUrl)
+				{
+					GlobalFree(ieConfig.lpszAutoConfigUrl);
+				}
+				if (ieConfig.lpszProxy)
+				{
+					GlobalFree(ieConfig.lpszProxy);
+				}
+				if (ieConfig.lpszProxyBypass)
+				{
+					GlobalFree(ieConfig.lpszProxyBypass);
+				}
+			}
+
+			// mark as "configured" so we don't attempt to do this horrible PoS mess again.
+			ctx->proxy_configured = TRUE;
+		}
+
+		if (ctx->proxy_for_url &&
+			!WinHttpSetOption(hReq, WINHTTP_OPTION_PROXY, ctx->proxy_for_url, sizeof(WINHTTP_PROXY_INFO)))
+		{
+			dprintf("[%s] Unable to set proxy options: %u", GetLastError());
+		}
+	}
+	else if (ctx->proxy_user)
 	{
 		dprintf("[%s] Setting proxy username to %S", direction, ctx->proxy_user);
 		dprintf("[%s] Setting proxy password to %S", direction, ctx->proxy_pass);
```

Comment:
Semantically there has to be one,  as you can't set a user without the
password and have it work properly.

Technically the user could pass just the user name, so a null check is
probably a good idea!